src/engine/Join.cpp

```diff
@@ -582,89 +734,142 @@ void updateRuntimeInfoForLazyScan(
-  auto rightBlocks = convertGenerator(std::move(rightBlocksInternal));
-
-  runtimeInfo().addDetail("time-for-filtering-blocks", timer.msecs());
-
-  auto doJoin = [&rowAdder](auto& left, auto& right) mutable {
-    ad_utility::zipperJoinForBlocksWithoutUndef(left, right, std::less{},
-                                                rowAdder);
-  };
-  auto blockForIdTable = std::span{&permutationIdTable, 1};
-  if (idTableIsRightInput) {
-    doJoin(rightBlocks, blockForIdTable);
-  } else {
-    doJoin(blockForIdTable, rightBlocks);
-  }
-  auto result = std::move(rowAdder).resultTable();
-  result.setColumnSubset(joinColMap.permutationResult());
+  auto resultPermutation = joinColMap.permutationResult();
+  return createResult(
+      requestLaziness,
+      [this, joinColTable, scan = std::move(scan), joinColScan,
+       resultWithIdTable = std::move(resultWithIdTable),
+       joinColMap = std::move(joinColMap)](
+          std::function<void(IdTable&, LocalVocab&)> yieldTable) {
+        const IdTable& idTable = resultWithIdTable->idTable();
+        ad_utility::AddCombinedRowToIdTable rowAdder{
+            1, IdTable{getResultWidth(), getExecutionContext()->getAllocator()},
+            cancellationHandle_, CHUNK_SIZE, std::move(yieldTable)};
+
+        AD_CORRECTNESS_CHECK(joinColScan == 0);
+        auto permutationIdTable = ad_utility::IdTableAndFirstCol{
+            idTable.asColumnSubsetView(idTableIsRightInput
+                                           ? joinColMap.permutationRight()
+                                           : joinColMap.permutationLeft()),
+            resultWithIdTable->getCopyOfLocalVocab()};
+
+        ad_utility::Timer timer{
+            ad_utility::timer::Timer::InitialStatus::Started};
+        bool idTableHasUndef =
+            !idTable.empty() && idTable.at(0, joinColTable).isUndefined();
+        std::optional<std::shared_ptr<const Result>> indexScanResult =
+            std::nullopt;
+        using FirstColView = ad_utility::IdTableAndFirstCol<IdTable>;
+        using GenWithDetails =
+            cppcoro::generator<FirstColView,
+                               CompressedRelationReader::LazyScanMetadata>;
+        auto rightBlocks = [&scan, idTableHasUndef, &permutationIdTable,
+                            &indexScanResult]()
+            -> std::variant<cppcoro::generator<FirstColView>, GenWithDetails> {
+          if (idTableHasUndef) {
+            indexScanResult =
+                scan->getResult(false, ComputationMode::LAZY_IF_SUPPORTED);
+            AD_CORRECTNESS_CHECK(
+                !indexScanResult.value()->isFullyMaterialized());
+            return convertGenerator(
+                std::move(indexScanResult.value()->idTables()));
+          } else {
+            auto rightBlocksInternal =
+                scan->lazyScanForJoinOfColumnWithScan(permutationIdTable.col());
+            return convertGenerator(std::move(rightBlocksInternal));
+          }
+        }();
```

Comment:
It's not that simple unfortunately because the lifetime of `indexScanResult` needs to be extended past the scope.
I'm sure there's a nicer solution for this, but this function needs to be changed anyway when implementing a block prefiltering for lazy results.